virt-handler, Add infoSource field to vmi status

oshoval opened the pull request:
**What this PR does / why we need it**:

In case of MAC change in the guest,
the vmi status will present two interfaces,
one based on the domain spec, and one based on the
guest agent (if it exists).

The reason is that the MAC address is used in order
to match the interface of the domain spec with the
interfaces reported by the guest agent.
Once the MAC is changed, this matching can't be performed.
   
In order to make the info easier to understand,
add a new field to each interface in vmi.status.
infoSource field has 3 possible values:
1. domain - info is based on the domain spec
2. guest-agent - info is based on guest agent report
3. domain, guest-agent - info is based on both domain spec
    and guest agent report.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=2013455

**Special notes for your reviewer**:
User Guide PR: https://github.com/kubevirt/user-guide/pull/504

**Release note**:
```release-note
Add infoSource field to vmi.status.interfaces.
The field specifies the origin of the interface data collected. values: domain, guest-agent, or both.
```

oshoval commented:
/uncc @jean-edouard @yuhaohaoyu

oshoval commented:
/test pull-kubevirt-check-unassigned-tests

oshoval commented:
/test pull-kubevirt-unit-test

oshoval commented:
/cc @EdDev

maiqueb commented:
> In case of MAC change in the guest,
the vmi status will present two interfaces,
one based on the domain spec, and one based on the
guest agent (if it exists).

Could you paste some examples of what the status representation will look like, so its easier to visualize ? As is, I am unsure I understand what you're proposing.

oshoval commented:
> > In case of MAC change in the guest,
> > the vmi status will present two interfaces,
> > one based on the domain spec, and one based on the
> > guest agent (if it exists).
> 
> Could you paste some examples of what the status representation will look like, so its easier to visualize ? As is, I am unsure I understand what you're proposing.

An interface where the domain and guest-agent both know to interface (usually the case)

```
  - infoSource: domain, guest-agent
    interfaceName: eth1
    ipAddress: 10.200.0.1
    ipAddresses:
    - 10.200.0.1
    - fe80::1e:d6ff:fe00:6b
    mac: 02:1e:d6:00:00:6b
    name: br10-vlan100-network
```

Now if you change the mac of the interface the situation will be
```
  - mac: 02:1e:d6:00:00:6b
    name: br10-vlan100-network
  - interfaceName: eth1
    ipAddress: 10.200.0.1
    ipAddresses:
    - 10.200.0.1
    - fe80::1e:d6ff:fe00:6b
    mac: 02:00:b5:b5:b5:c9
```

This PR suggest to add the infoSource field as follows

```
  - infoSource: domain
    mac: 02:1e:d6:00:00:6b
    name: br10-vlan100-network
  - infoSource: guest-agent
    interfaceName: eth1
    ipAddress: 10.200.0.1
    ipAddresses:
    - 10.200.0.1
    - fe80::1e:d6ff:fe00:6b
    mac: 02:00:b5:b5:b5:c9
```

maiqueb commented:
```yaml
  - infoSource: domain
    mac: 02:1e:d6:00:00:6b
    name: br10-vlan100-network
  - infoSource: guest-agent
    interfaceName: eth1
    ipAddress: 10.200.0.1
    ipAddresses:
    - 10.200.0.1
    - fe80::1e:d6ff:fe00:6b
    mac: 02:00:b5:b5:b5:c9
```
But do the above 2 entries refer to the same entity ? From your PR's description, I think so.

How can I - or anything that reads the interface status - infer that @oshoval ? 

AFAIU, there isn't any attribute there to map them together.

oshoval commented:
> ```yaml
>   - infoSource: domain
>     mac: 02:1e:d6:00:00:6b
>     name: br10-vlan100-network
>   - infoSource: guest-agent
>     interfaceName: eth1
>     ipAddress: 10.200.0.1
>     ipAddresses:
>     - 10.200.0.1
>     - fe80::1e:d6ff:fe00:6b
>     mac: 02:00:b5:b5:b5:c9
> ```
> 
> But do the above 2 entries refer to the same entity ? From your PR's description, I think so.
> 
> How can I - or anything that reads the interface status - infer that @oshoval ?
> 
> AFAIU, there isn't any attribute there to map them together. Probably if you add a more complete example I'll understand.

We can't know if they refer to the same entry (and this PR doesnt meant to do that),
once the mac is changed, there is no way to know that.
Its like creating a new interface once you change the mac (mac was the glue and its broken now as the PR desc says).
it only shows that the data in the spec now has two sources:
1. the virsh domain xml 
2. the guest agent report

maiqueb commented:
> > ```yaml
> >   - infoSource: domain
> >     mac: 02:1e:d6:00:00:6b
> >     name: br10-vlan100-network
> >   - infoSource: guest-agent
> >     interfaceName: eth1
> >     ipAddress: 10.200.0.1
> >     ipAddresses:
> >     - 10.200.0.1
> >     - fe80::1e:d6ff:fe00:6b
> >     mac: 02:00:b5:b5:b5:c9
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But do the above 2 entries refer to the same entity ? From your PR's description, I think so.
> > How can I - or anything that reads the interface status - infer that @oshoval ?
> > AFAIU, there isn't any attribute there to map them together. Probably if you add a more complete example I'll understand.
> 
> No, we can't know if they refer to the same entry (and this PR doesnt meant to do that), once the mac is changed, there is no way to know that. Its like creating a new interface once you change the mac (mac was the glue and its broken now as the PR desc says). it only shows that the data in the spec now has two sources:
> 
> 1. the virsh domain xml
> 2. the guest agent report
> 
> It will be easier to know what in the domain.spec and what actually exists in the guest (and reported by the agent).

But what can I do with this information ? I know the current state of **the** interface, and I know what the original state of **an** interface on the system was. But I got no idea as to which one. The latter hardly seems useful.

While I'm aware this fixes a bug, that `domain` information in the status is very confusing to me.

oshoval commented:
> > > ```yaml
> > >   - infoSource: domain
> > >     mac: 02:1e:d6:00:00:6b
> > >     name: br10-vlan100-network
> > >   - infoSource: guest-agent
> > >     interfaceName: eth1
> > >     ipAddress: 10.200.0.1
> > >     ipAddresses:
> > >     - 10.200.0.1
> > >     - fe80::1e:d6ff:fe00:6b
> > >     mac: 02:00:b5:b5:b5:c9
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > But do the above 2 entries refer to the same entity ? From your PR's description, I think so.
> > > How can I - or anything that reads the interface status - infer that @oshoval ?
> > > AFAIU, there isn't any attribute there to map them together. Probably if you add a more complete example I'll understand.
> > 
> > 
> > No, we can't know if they refer to the same entry (and this PR doesnt meant to do that), once the mac is changed, there is no way to know that. Its like creating a new interface once you change the mac (mac was the glue and its broken now as the PR desc says). it only shows that the data in the spec now has two sources:
> > 
> > 1. the virsh domain xml
> > 2. the guest agent report
> > 
> > It will be easier to know what in the domain.spec and what actually exists in the guest (and reported by the agent).
> 
> But what can I do with this information ? I know the current state of **the** interface, and I know what the original state of **an** interface on the system was. But I got no idea as to which one. The latter hardly seems useful.
> 
> While I'm aware this fixes a bug, that `domain` information in the status is very confusing to me.

Think from operator perspective, it will show you at once the actual and configured,
instead of going to logs and virsh dumpxml.
It will also show you if there is discrepancy between the domain spec and guest agent report.
And this is also a private case, i mean that if you manually add a new interface, it will show you that it appears only on "guest-agent".
Before guest agent report, you will see source as "domain", it will also tell you what is the source of the information, and that the guest-agent didn't report yet (unless it doesn't exists / broken and then it won't report, and it will stay as "domain").

maiqueb commented:
> > > > ```yaml
> > > >   - infoSource: domain
> > > >     mac: 02:1e:d6:00:00:6b
> > > >     name: br10-vlan100-network
> > > >   - infoSource: guest-agent
> > > >     interfaceName: eth1
> > > >     ipAddress: 10.200.0.1
> > > >     ipAddresses:
> > > >     - 10.200.0.1
> > > >     - fe80::1e:d6ff:fe00:6b
> > > >     mac: 02:00:b5:b5:b5:c9
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > But do the above 2 entries refer to the same entity ? From your PR's description, I think so.
> > > > How can I - or anything that reads the interface status - infer that @oshoval ?
> > > > AFAIU, there isn't any attribute there to map them together. Probably if you add a more complete example I'll understand.
> > > 
> > > 
> > > No, we can't know if they refer to the same entry (and this PR doesnt meant to do that), once the mac is changed, there is no way to know that. Its like creating a new interface once you change the mac (mac was the glue and its broken now as the PR desc says). it only shows that the data in the spec now has two sources:
> > > 
> > > 1. the virsh domain xml
> > > 2. the guest agent report
> > > 
> > > It will be easier to know what in the domain.spec and what actually exists in the guest (and reported by the agent).
> > 
> > 
> > But what can I do with this information ? I know the current state of **the** interface, and I know what the original state of **an** interface on the system was. But I got no idea as to which one. The latter hardly seems useful.
> > While I'm aware this fixes a bug, that `domain` information in the status is very confusing to me.
> 
> Think from operator perspective, it will show you at once the actual and configured, instead of going to logs and virsh dumpxml. 

But I already had - and still have - the `configured` value: in the spec. 

The status should have information about the current value. I feel like the API is drifting from its purpose.

oshoval commented:
Please don't quote all the history if possible.

> But I already had - and still have - the `configured` value: in the spec.

see please the example where you don't configure a MAC in the vmi.spec (and no KMP active)
the MAC is configured, but not in the vmi.spec , but in the vmi.status

> The status should have information about the current value. I feel like the API is drifting from its purpose.

Lets see please other opinions about this

maiqueb commented:
> > But I already had - and still have - the `configured` value: in the spec.
> 
> see please the example where you don't configure a MAC in the vmi.spec (and no KMP active) the MAC is configured, but not in the vmi.spec , but in the vmi.status

So ? Why would I care about what was the interface's MAC address when it was created ? I don't see what this changes.

EdDev commented:
> So ? Why would I care about what was the interface's MAC address when it was created ? I don't see what this changes.

@maiqueb there are 3 states: The desired (VMI spec), the domain state (domxml) and guest agent visibility.
The domxml represent an actual state and not a desired state, at least in the virt perspective. The guest agent state is showing a different layer, which may or may not be in sync with the actual virt/emulated devices the domain provides.

I think this is also playing along with the option you are working on these days... a device may be added, and it may or may not succeed (domain state), and it may or may not appear in the guest (because the user decided to remove the network driver from it?).

We are not talking about the desired state taking affect on the reported information. We report domain and guest-agent information with some pod info.

Does this make sense now?

maiqueb commented:
> > So ? Why would I care about what was the interface's MAC address when it was created ? I don't see what this changes.
> 
> @maiqueb there are 3 states: The desired (VMI spec), the domain state (domxml) and guest agent visibility. The domxml represent an actual state and not a desired state, at least in the virt perspective. The guest agent state is showing a different layer, which may or may not be in sync with the actual virt/emulated devices the domain provides.

I am only concerned about the API here. Which **lives forever**. The domxml is an implementation detail. In regards to the API, I couldn't care less about it. 

> 
> I think this is also playing along with the option you are working on these days... a device may be added, and it may or may not succeed (domain state), and it may or may not appear in the guest (because the user decided to remove the network driver from it?).

Since you mentioned it, I got a question about hot-unplug: imagine I hot-unplug the interface. How can I **ever** remove the domain interface from the VMI status ? I have **no idea** whatsoever which of the entries it is - i.e. it is missing the interface name:

```yaml
- infoSource: domain
  mac: 02:1e:d6:00:00:6b
  name: br10-vlan100-network
```

> 
> We are not talking about the desired state taking affect on the reported information. We report domain and guest-agent information with some pod info.

I understand what you do. But: why would the user want the domain information ? That's what I'm missing. As I see it, whoever uses the VMI state is interested in one thing: what's the current VM's MAC.

> 
> Does this make sense now?

First, thanks for taking the time to write this articulated explanatory answer. I understand the reasons for this design. 

... But it's not about making sense, it's about providing meaningful information to the user that can be acted upon; this solution, as I see it not only breaks the kubernetes API model (you knowingly show stuff in the status that is not up to date) but is also confusing, and can't be mapped to anything.

Having said that, sometimes the rules must be broken to achieve XYZ. It might be OK that the status shows information that is out dated and / or confusing. What I am missing is why / what for would a user be interested on the domain defined MAC address.

EdDev commented:
> > > So ? Why would I care about what was the interface's MAC address when it was created ? I don't see what this changes.
> > 
> > 
> > @maiqueb there are 3 states: The desired (VMI spec), the domain state (domxml) and guest agent visibility. The domxml represent an actual state and not a desired state, at least in the virt perspective. The guest agent state is showing a different layer, which may or may not be in sync with the actual virt/emulated devices the domain provides.
> 
> I am only concerned about the API here. Which **lives forever**. The domxml is an implementation detail. In regards to the API, I couldn't care less about it.

The domxml represent the real domain created, which tells us if there is a device plugged in or not.
So I do not understand how this fact relates to an implementation detail here. Whatever the implementation is, if it is defined in the domain or not is of value.

Are you claiming here that if we have a guest application like guest-agent that does not report an interface, it means that this is the bottom line fact? I fail to understand the logic here.

> 
> > I think this is also playing along with the option you are working on these days... a device may be added, and it may or may not succeed (domain state), and it may or may not appear in the guest (because the user decided to remove the network driver from it?).
> 
> Since you mentioned it, I got a question about hot-unplug: imagine I hot-unplug the interface. How can I **ever** remove the domain interface from the VMI status ? I have **no idea** whatsoever which of the entries it is - i.e. it is missing the interface name:
> 
> ```yaml
> - infoSource: domain
>   mac: 02:1e:d6:00:00:6b
>   name: br10-vlan100-network
> ```

I do not understand how this is related to the topic we are discussing now.
But you are suppose to detach an interface from the VMI Spec, not from the status. And there you may have a problem to distinguish between a list of interfaces indeed, just because the current API (in the VMI spec) not that ideal (to say the least).

> 
> > We are not talking about the desired state taking affect on the reported information. We report domain and guest-agent information with some pod info.
> 
> I understand what you do. But: why would the user want the domain information ? That's what I'm missing. As I see it, whoever uses the VMI state is interested in one thing: what's the current VM's MAC.

I think the operator (which sees the VMI Status) is interested to see as much information as possible about the domain/guest that runs there. That means, to understand if the domain is plugged with the vNIC or not, and if it is, what is seen in the guest (in case the guest agent is alive).
You may claim that the operator cannot trust the guest-agent, which is completely correct. But this is exactly why we tell the operator something like this: The device is there, but the agent in the guest is not reporting it.
Now the operator can tell the user of the guest, that the problem is in his backyard and not in the domain that was wired to him.

> 
> > Does this make sense now?
> 
> First, thanks for taking the time to write this articulated explanatory answer. I understand the reasons for this design.
> 
> ... But it's not about making sense, it's about providing meaningful information to the user that can be acted upon; this solution, as I see it not only breaks the kubernetes API model (you knowingly show stuff in the status that is not up to date) but is also confusing, and can't be mapped to anything.

My claim is that until now we lied to the operator in our status report and now we provide him with the most accurate and transparent information we can. I do not see how we break any K8S API model, we report reality.

> Having said that, sometimes the rules must be broken to achieve XYZ. It might be OK that the status shows information that is out dated and / or confusing. What I am missing is why / what for would a user be interested on the domain defined MAC address.

The user in this case is the operator, and he wants to know that what he requested in the desired state, is actually applied.
If the user of the guest, decided to override it, the operator should know because:
- In the attempt to report the status, we depend on the MAC to provide a mapping to the network name (which is also the interface name in the spec, unfortunately). Therefore, if we cannot do it reliably, we need to give the "why", or better, to be transparent on what is going on on the system.
- The traffic may not work, because mac-spoofing is enabled, and it is the fault of the guest user and not of the operator. The operator will be able to conclude this based on the info we provide in the status.

oshoval commented:
rebase

oshoval commented:
/test pull-kubevirt-goveralls 
/test pull-kubevirt-build 

something is weird with pull-kubevirt-verify-go-mod  (even that its not required, at least with the label) on it

oshoval commented:
/retest

EdDev commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -269,12 +270,20 @@ func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceSt
 		}
 	}
+
+	for i, _ := range interfaceStatuses {
```
nit: You do not need the second variable `_`.

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,101 @@
+			vmiSpec := libvmi.NewFedora(
+				libvmi.WithInterface(*libvmi.InterfaceWithMac(&bridgeInterface, originalMac)),
+				libvmi.WithNetwork(kvirtv1.DefaultPodNetwork()))
+			tests.AddUserData(vmiSpec, "cloud-init", "#!/bin/bash\nsudo ip link set dev eth0 address "+updatedMac+"\n")
```
There is a dedicated build function for this: `WithCloudInitNoCloudUserData`. I did not work for you?

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,101 @@
+
+		BeforeEach(func() {
+			bridgeInterface := libvmi.InterfaceDeviceWithBridgeBinding()
+			vmiSpec := libvmi.NewFedora(
```
Will this work with Cirros?

oshoval replied:
we don't have guest agent there, correct me if i am wrong

oshoval replied:
as we spoke, we need the guest-agent here

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,101 @@
+				vmiObj, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+
+				mac, err := getInterfaceMacByInfoSource(vmiObj, "guest-agent")
+				if err != nil {
+					return err
+				}
+				Expect(mac).To(Equal(updatedMac))
+
+				mac, err = getInterfaceMacByInfoSource(vmiObj, "domain")
+				if err != nil {
+					return err
+				}
+				Expect(mac).To(Equal(originalMac))
```
I prefer to see an assert on the full list of interfaces reported, not individual ones. This way we will get a full picture if something fails.

Also, if any of the asserts you just added fail, there will be no retry anymore. Was this your intention?

oshoval replied:
will change it to the full list

yes this was the intention
because once the guest-agent appears alone, it means that from this point the other must be domain alone
(it is implicit), even if for example in the beginning there was a non conflict match

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,101 @@
+	"kubevirt.io/kubevirt/tests/util"
+)
+
+var _ = SIGDescribe("Infosource", func() {
```
I think you should test all possible scenarios, which include:
- mac changed (the current covered scenario).
- missing interface in the guest.
- an additional interface in the guest.
- an interface which is in sync (domain + guest-agent).

This can still be a single test, but with multiple interfaces.

oshoval replied:
bullet 2 and 3 are covered by the first bullet,
as once you change the MAC you get an interface which appear only in the domain (bullet 2),
and an interface that appears only on the guest (bullet 3)

I will add bullet 4 (planned to)

EdDev replied:
I think that the outcome is the same, but not the scenario to begin with.
Several scenarios may reach the same outcome, but it makes sense to test them all to make sure we are not missing anything from a user perspective.

oshoval replied:
added 2 + 3

EdDev reviewed (changes requested):
It will probably take me a few days to decrypt `updateInterfacesFromDomain`. it is overwhelming to be sure all is ok and that all options have been covered. Even the `agent_parser` unit tests are very hard to understand.

oshoval commented:
addressed comments

oshoval commented:
/test pull-kubevirt-generate

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,146 @@
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	k8snetworkplumbingwgv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
```
Can you call it something shorter? Maybe `k8snadclientv1`? (or just `nadv1`)

oshoval replied:
done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,146 @@
+	})
+})
+
+func generateNetworkAttachmentDefinition() *k8snetworkplumbingwgv1.NetworkAttachmentDefinition {
```
We usually use the term `new` and not `generate` for returning structures like these.

oshoval replied:
done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,146 @@
+
+		BeforeEach(func() {
+			By("Create NetworkAttachmentDefinition")
+			nad := generateNetworkAttachmentDefinition()
+			_, err := virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(util.NamespaceTestDefault).Create(context.TODO(), nad, metav1.CreateOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
```
@orelmisan , you have raised interest to use something like this in the multus tests. It seems this package was already used, so maybe it will fit to what you need as well.

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,146 @@
+				},
+			}
+
+			Eventually(func() error {
```
We usually put inside the `Eventually` the data collection and early assertions, but do not embed it with the final assertion.
Can this be split here as well?

oshoval replied:
done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,146 @@
+				Expect(err).NotTo(HaveOccurred())
+
+				if len(vmiObj.Status.Interfaces) != len(expectedInterfaces) {
+					return fmt.Errorf("interfaces doesn't match len error")
```
I think you need to be consistent and return the same error as the one on line 133.

oshoval replied:
done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,146 @@
+func generateNetworkAttachmentDefinition() *k8snetworkplumbingwgv1.NetworkAttachmentDefinition {
+	return &k8snetworkplumbingwgv1.NetworkAttachmentDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "secondary",
```
This is too generic name, it may be used by other tests and then it will cause a mess.

oshoval replied:
done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,146 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "secondary",
+		},
+		Spec: k8snetworkplumbingwgv1.NetworkAttachmentDefinitionSpec{Config: `{"cniVersion": "0.3.1", "name": "secondary", "type": "cnv-bridge", "bridge": "br1"}`},
```
Same here, the `br1` is too generic, try to give it a more specialized name of you need an unique bridge/nad.

If there is no need for an unique bridge/nad, then consider showing that you share the same one that is already used under multus tests.
Also, please make sure we can run such tests in parallel, as it does use a common resource on the node. Not sure if it is a problem or not, it just needs to be considered well so we will not cause collisions.

oshoval replied:
> Also, please make sure we can run such tests in parallel, as it does use a common resource on the node. Not sure if it is a problem or not, it just needs to be considered well so we will not cause collisions.

It is not a problem (and was tested).
It is created in a namespace which is parallel safe,
each parallel instance creates its own namespaces
```
			_, err := virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(util.NamespaceTestDefault).Create(context.TODO(), nad, metav1.CreateOptions{})
```

I think the code here has only what we need, and the creation flow is simpler than
https://github.com/kubevirt/kubevirt/blob/77193f17822e4e3a781c116d340a7cbb7e42c6ce/tests/network/vmi_multus.go#L162
which uses POST, and more unneeded variables.

so i prefer to use this,
we can consider a follow PR that will create lib for NADs and will be used on both places
but for now the code on multus is more complex than what we need

EdDev replied:
> It is not a problem (and was tested).
> It is created in a namespace which is parallel safe,
> each parallel instance creates its own namespaces, and then we use it as the NAD ns

I do not know how you are able to be sure it is not a problem when running in parallel. It is not like you control which tests run at the same time. I cannot think of a reasonable way to control this.

I pointed out the bridge created on the node and you are talking about the NAD which can be consumed by pods.
I think you miss the part where this CNI bridge creates in a lazy manner a linux-bridge on the node. That part is a shared resource and has nothing to do with the namespace. If I am wrong, please correct me.

> so i prefer to use this, as its clean and compact (and uses the basic high level API which suitable here)

I do not mind you using the structured version, I do mind that two NAD/s point to the same underlying bridge, that sounds problematic to me.

oshoval replied:
> I do not know how you are able to be sure it is not a problem when running in parallel. It is not like you control which tests run at the same time. I cannot think of a reasonable way to control this.

In aspect NADs only, we do control the namespaces, each parallel instance has its own namespaces.

about the underlying bridge that is created on the node (rest of the comment), i changed the name of the bridge to `infosrc`, would it be enough ?

oshoval commented on pkg/virt-handler/vm.go:
```diff
@@ -873,9 +873,10 @@ func (d *VirtualMachineController) updateInterfacesFromDomain(vmi *v1.VirtualMac
+					Name:       network.Name,
+					IP:         podIface.PodIP,
+					IPs:        podIface.PodIPs,
+					InfoSource: "domain",
```
> This is interesting. I do not think the source of this information is the domain, it seems to be the pod itself.

Right, but i think we should have only guest-agent, and non guest-agent resolution

oshoval replied:
about bullet 2, moved it from the starting block to the end of the main loop as we spoke,
it is coming from the pod, but also from the domain, but when it coming from the domain,
we are not going into details that it some info is also from the pod (leaving it as internal details).
moving it to the main loop is less confusing.

oshoval replied:
about 1, defined consts

oshoval commented:
rebase

oshoval commented:
Thanks, addressed comments
leftovers - change to const (waiting for https://github.com/kubevirt/kubevirt/pull/6874, will rebase on top of it when it is lgtmed)

Checked manually what happens if guest-agent was stopped after its first report,
the whole report stays the same.
The reason is there are two caches that take effect here:
1. virt-launcher guest-agent report cache which saves the info that will be used when virt-handler request domain update every 5 mintues.
2. virt-handler domain cache, which is the domain that is used on VmiSync, and is updated when virt-launcher sends domain update (either pushed by virt-launcher when there are updates or polled every 5 minutes)

Talked with Meni, Meni is fine that it will be checked as tier 2 only, since its a long test and more suitable to tier 2 imho.

oshoval commented:
/test pull-kubevirt-goveralls

EdDev commented:
> Checked manually what happens if guest-agent was stopped after its first report, `sudo systemctl stop qemu-guest-agent.service` the whole report stays the same. The reason is there are two caches that take effect here:
> 
>     1. virt-launcher guest-agent report cache which saves the info that will be used when virt-handler request domain update every 5 mintues.
> 
>     2. virt-handler domain cache, which is the domain that is used on VmiSync, and is updated when virt-launcher sends domain update (either pushed by virt-launcher when there are updates or polled every 5 minutes)

I do not understand to which 2nd cache you refer here.
 
> Talked with Meni, Meni is fine that it will be checked as tier 2 only, since its a long test and more suitable to tier 2 imho.

Who is Meni and what is this tier 2?

Stopping the guest-agent and understanding its implications can be taken out of this PR, but I would expect to see text describing the scenario for future us.

EdDev commented:
@oshoval , please go over each comment you treated and either mention "Done" or elaborate what you did so I can track it.
It is inconvenient to resolve the comment without the reviewer seeing what was done and accepting it.

oshoval commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -253,6 +253,7 @@ func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceSt
 	for i, interfaceStatus := range interfaceStatuses {
 		if alias, exists := aliasByMac[interfaceStatus.Mac]; exists {
 			interfaceStatuses[i].Name = alias
+			interfaceStatuses[i].InfoSource = "domain, guest-agent"
```
Removed comment

oshoval commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -269,12 +270,20 @@ func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceSt
+		if interfaceStatuses[i].InfoSource == "" {
+			interfaceStatuses[i].InfoSource = "guest-agent"
+		}
+	}
```
moved as else

oshoval commented:
> @oshoval , please go over each comment you treated and either mention "Done" or elaborate what you did so I can track it. It is inconvenient to resolve the comment without the reviewer seeing what was done and accepting it.

done

oshoval commented:
> Who is Meni and what is this tier 2?
Meni Yakove, cnv-tests

I did explain, I will try to elaborate with the interesting parts (i am omitting cache 2 of virt-handler because it is less important to the following explanation)

Once you stop the guest-agent, the guest-agent poller on virt-launcher would stop generating `agentUpdate` events.
The `interfaceStatuses` variable [1] of virt-launcher domain notifier would not be changed, and preserve the last value,
so even other events would still report the same interfaces that were report last.

[1] https://github.com/kubevirt/kubevirt/blob/debf29e7bf0df011248b3662c46bfa55cf3f6750/pkg/virt-launcher/notify-client/client.go#L400

Each 300 seconds (virt-handler `defaultDomainResyncPeriodSeconds`),
virt-handler polls virt-launcher domain update which triggers this function on virt-launcher
https://github.com/kubevirt/kubevirt/blob/0119f397a54f4ac0bc585a42cf1395e732c3295f/pkg/virt-launcher/virtwrap/cmd-server/server.go#L350
`GetDomain` calls `InterfacesStatus` which use the `AsyncAgentStore` which also have the raw guest-agent interface report.

Those are all the places that virt-launcher reports the domain, and since all of them have caching,
the virt-handler which receives the domain, will always get the same data of the last time guest-agent did report.
Hence the fact that guest agent is stopped won't be reflected.

Note that if there is a guest-agent running, it reports interfaces once in 2 minutes in the steady state,
so those caches are important as long as it runs, so the status.interfaces won't be flickering (comes and goes).

oshoval commented:
> Who is Meni and what is this tier 2?

Meni Yakove, cnv-tests

I did explain, I will try to elaborate with the interesting parts (i am omitting the 2nd cache of virt-handler because it is less important to the following explanation)

Once you stop the guest-agent, the guest-agent poller on virt-launcher would stop generating `agentUpdate` events.
The `interfaceStatuses` variable (see following link) of virt-launcher domain notifier would not be changed, and preserve the last value,
so even other events (non guest agent events) would still report the same interfaces that were report last.

https://github.com/kubevirt/kubevirt/blob/debf29e7bf0df011248b3662c46bfa55cf3f6750/pkg/virt-launcher/notify-client/client.go#L400

Each 300 seconds (virt-handler `defaultDomainResyncPeriodSeconds`),
virt-handler polls virt-launcher domain update which triggers this function on virt-launcher
https://github.com/kubevirt/kubevirt/blob/0119f397a54f4ac0bc585a42cf1395e732c3295f/pkg/virt-launcher/virtwrap/cmd-server/server.go#L350
`GetDomain` calls `InterfacesStatus` which use the `AsyncAgentStore` which also have the raw guest-agent interface report.

Those are all the places that virt-launcher reports the domain, and since all of them have caching (as explained above),
the virt-handler which receives the domain, will always get the same data of the last time guest-agent did report.
Hence the fact that guest agent is stopped won't be reflected in the `status.interfaces`.

Note that if there is a guest-agent running, it reports interfaces once in 2 minutes in the steady state,
so those caches are important as long as it runs, so the status.interfaces won't be flickering (comes and goes).

oshoval commented:
1. rebase main
2. rebase on top of https://github.com/kubevirt/kubevirt/pull/6874
(so i can add the const to the file it presents pkg/network/vmispec/interface.go)

/hold 
due to 2

oshoval commented:
defined consts

oshoval commented:
rebase

/hold cancel

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -955,14 +955,18 @@ func (d *VirtualMachineController) updateInterfacesFromDomain(vmi *v1.VirtualMac
 				}
 				delete(domainInterfaceStatusByMac, interfaceMAC)
+			} else {
+				newInterface.InfoSource = netvmispec.InfoSourceDomain
```
This will work for regular interfaces but not for SR-IOV interfaces.
Can you please add a check on the SR-IOV e2e tests with the expected value?

oshoval replied:
> This will work for regular interfaces but not for SR-IOV interfaces.

for the sriov it will add the interfaces from 
https://github.com/kubevirt/kubevirt/pull/6793/files#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR984
because its a hosdev and not in the `domain.Spec.Devices.Interfaces`

EdDev replied:
What about cases where you have no guest-agent? How will we see the SR-IOV interface?
Can it be that we do not report it today at all if no guest-agent is present? It doesn't seem right, I think it is reported through the pod interfaces. Line 866 above.
This requires checking/testing.

oshoval replied:
we don't report them if there is no guest-agent, its not new
you can also see that before the first guest agent report we don't see them

Lets say this is the yaml interfaces / network
```
      interfaces:
      - masquerade: {}
        name: default
      - macAddress: 02:00:00:e1:62:a2
        name: sriov3
        sriov: {}


  networks:
  - name: default
    pod: {}
  - multus:
      networkName: sriov
    name: sriov3
```

Before first guest agent report, only the pod network is added via line 866 (as it looks on the interfaces from the pod)
```
  interfaces:
  - infoSource: domain
    ipAddress: 10.244.1.8
    ipAddresses:
    - 10.244.1.8
    mac: 52:54:00:bc:ad:c6
    name: default
```

After the guest agent report, we see all the info,
we can still see that sriov is only from the guest-agent according infoSource,
this is why there is the code block
https://github.com/kubevirt/kubevirt/pull/6793/files#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR973-R984
which adds all the interfaces that were not handled before in this function (hostdev that are not in the domain.spec such as sriov, manually added interfaces) 
```
  - infoSource: domain, guest-agent
    interfaceName: eth0
    ipAddress: 10.244.1.8
    ipAddresses:
    - 10.244.1.8
    mac: 52:54:00:bc:ad:c6
    name: default
  - infoSource: guest-agent
    interfaceName: sriov
    ipAddress: 192.168.1.1
    ipAddresses:
    - 192.168.1.1
    - fe80::ff:fee1:62a2
    mac: 02:00:00:e1:62:a2
    name: sriov3
```

oshoval replied:
The above snippets are from a real run (there was a guest agent installed. but before the first report this what was shown).
The question on stack is if `getPodInterfacefromFileCache` have something about the sriov data when there is no guest agent (or when there is unless i am wrong in the observation)
https://github.com/kubevirt/kubevirt/blob/7655a29a3d010234f56c5ae084c13c07b1033eb2/pkg/virt-handler/vm.go#L870
anyhow imho it is not related to this PR, and we can take it offline, unless you think otherwise

oshoval replied:
In the scenario I ran there were 2 interfaces, imo it affects the flow.
The problem is not new, i am aware of it and described it above, but np lets take it to a follow effort please.

EdDev reviewed (changes requested):
Passed over all production and unit tests.
Still reviewing the e2e tests.

The only major issue I see is with the SR-IOV, see inline comment.

oshoval commented on pkg/virt-handler/vm.go:
```diff
@@ -955,14 +955,18 @@ func (d *VirtualMachineController) updateInterfacesFromDomain(vmi *v1.VirtualMac
 			// Remove the interface from domainInterfaceStatusByMac to mark it as handled
 			if interfaceStatus, exists := domainInterfaceStatusByMac[interfaceMAC]; exists {
 				newInterface.InterfaceName = interfaceStatus.InterfaceName
+				newInterface.InfoSource = interfaceStatus.InfoSource
```
> If I understand correctly, this info may include the fact that the source is guest-agent only or both guest-agent and domain. Is this correct?

It will look on `domainInterfaceStatusByMac` (derived from `domain.Status.Interfaces`) and take InfoSource from the,
`InfoSource` was calculated on the merge function of virt-launcher (and can be either one of the 3 options)

EdDev replied:
> There are actually 3 places, because virt-launcher merge is called from two places in virt-launcher.
(the one where launcher push data, and the one that is called every 300 seconds via GetDomain)

Thank you for the clarification.
Another question: Is this data populated if there is no guest-agent?

oshoval replied:
> Another question: Is this data populated if there is no guest-agent?

Yes, the domain.spec will be populated, once in 300 seconds via GetDomain (virt-handler initiated the request every 300s), or upon every trigger of virt-launcher domain modify event (if there are any when there is no guest-agent data)

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+		})
+
+		It("should have the expected entries in vmi status", func() {
+			expectedInterfaces := []kvirtv1.VirtualMachineInstanceNetworkInterface{
+				{
+					InfoSource: netvmispec.InfoSourceDomain,
+					MAC:        originalMac,
+					Name:       defaultInterfaceName,
+				},
+				{
+					InfoSource:    netvmispec.InfoSourceDomainAndGA,
+					InterfaceName: "eth1",
+					MAC:           secondaryLinuxBridgeInterface1.MacAddress,
+					Name:          secondaryLinuxBridgeInterface1.Name,
+				},
+				{
+					InfoSource: netvmispec.InfoSourceDomain,
+					MAC:        secondaryLinuxBridgeInterface2.MacAddress,
+					Name:       secondaryLinuxBridgeInterface2.Name,
+				},
+				{
+					InfoSource:    netvmispec.InfoSourceGuestAgent,
+					InterfaceName: "eth0",
+					MAC:           updatedMac,
+				},
+				{
+					InfoSource:    netvmispec.InfoSourceGuestAgent,
+					InterfaceName: "dummy0",
+					MAC:           dummyInterfaceMac,
+				},
+			}
```
Can you split it in a way we can easily see what are the interfaces expected to be in sync (domain + guest) and the ones only on the domain or guest? Maybe it will not be needed if the const names are updated.

oshoval replied:
I will start with changing the names, we can't split it, unless we reorder the vmi.status.interfaces as well
The reason is that the ones which appear on the domain are ordered as their spec order when the vmi status is built,
Where the guest agent ones order is random (see https://github.com/kubevirt/kubevirt/pull/6793#discussion_r768765169)

Another thing i might be able to do is switch between the purpose of secondaryLinuxBridgeInterface1 and secondaryLinuxBridgeInterface2 , so it will be domain, domain, domainAndGa, guest-agent, guest-agent
instead domain, domainAndGa, domain, guest-agent, guest-agent

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+// which means there was a guest-agent report finally, and then we can compare the rest expected info.
+func dummyInterfaceExists(vmi *kvirtv1.VirtualMachineInstance) bool {
+	for i := range vmi.Status.Interfaces {
+		if vmi.Status.Interfaces[i].InterfaceName == "dummy0" {
```
Please use a const for the `dummy0`.

oshoval replied:
done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+				Expect(err).NotTo(HaveOccurred())
+
+				return dummyInterfaceExists(vmiObj)
+			}, 140*time.Second, 2*time.Second).Should(Equal(true))
```
Why do we need more than 2min after the guest-agent is up to see this?

oshoval replied:
it will stop before that, we need 2 minutes in the worst case (if we passed the transient exponential backoff of guest-agent polling).

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+				return dummyInterfaceExists(vmiObj)
+			}, 140*time.Second, 2*time.Second).Should(Equal(true))
+
+			Expect(len(vmiObj.Status.Interfaces)).To(Equal(len(expectedInterfaces)))
```
If this fails, we will not have any valuable information to examine.
Won't the rest of the assertion detect this as well?

oshoval replied:
> Won't the rest of the assertion detect this as well?

next line is copying ips from one array to another, they must be the same length, else we might get index out of bounds,
i will add dump of the interfaces in case there is an assert

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+}
+
+// canonizeInterfaceStatus orders "guest-agent" only records in predictable order
+func canonizeInterfaceStatus(vmi *kvirtv1.VirtualMachineInstance) {
```
Can you please explain why this can happen? I would expect the order to be consistent.
In any case, the name is not reflecting what it does really... you just order here the list.

Also, please pass in only what is needed (`Interfaces`) and make this immutable (i.e. return a new list).

oshoval replied:
> Can you please explain why this can happen? I would expect the order to be consistent.

Guest agent report them in a random order, even on the same run

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+					"#!/bin/bash\n"+changeMacCmd+
+						createDummyInterfaceCmd+
+						setnsCmd,
+					false))
```
done

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+			changeMacCmd := "sudo ip link set dev eth0 address " + updatedMac + "\n"
+			createDummyInterfaceCmd := "ip link add dummy0 type dummy\n" +
+				"ip link set dev dummy0 address " + dummyInterfaceMac + "\n"
+			setnsCmd := "ip netns add testns\n" +
```
done

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+			_, err := virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(util.NamespaceTestDefault).Create(context.TODO(), nad, metav1.CreateOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			changeMacCmd := "sudo ip link set dev eth0 address " + updatedMac + "\n"
```
for some reason i had to add sudo, but rechecked and it works without (will just verify manually that i see the change, but the test do work and it does expect the new mac)
(btw there are commands that do need sudo i think, it depends on the /etc/sudoers IIUC, we had this in the past)

updated the name

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+			By("Create NetworkAttachmentDefinition")
+			nad := newNetworkAttachmentDefinition(secondaryNetworkName)
+			_, err := virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(util.NamespaceTestDefault).Create(context.TODO(), nad, metav1.CreateOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
```
done

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+			},
+		}
+
+		secondaryLinuxBridgeInterface1 := kvirtv1.Interface{
```
done

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+		const dummyInterfaceMac = "02:00:b0:b0:b0:b0"
+		const secondaryNetworkName = "infosrc"
+
+		linuxBridgeNetwork1 := kvirtv1.Network{
```
added first commit which adds the network name to that function, and adapted to it

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+			Name: "bridge1-unchanged",
+			NetworkSource: kvirtv1.NetworkSource{
+				Multus: &kvirtv1.MultusNetwork{
+					NetworkName: secondaryNetworkName,
```
changed

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+		const originalMac = "02:00:05:05:05:05"
+		const updatedMac = "02:00:b5:b5:b5:b5"
+		const dummyInterfaceMac = "02:00:b0:b0:b0:b0"
+		const secondaryNetworkName = "infosrc"
```
done

oshoval commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,210 @@
+		tests.BeforeTestCleanup()
+	})
+
+	Context("with a vmi and a few interfaces", func() {
```
done

oshoval commented:
addressed comments

oshoval commented:
fix naming

oshoval commented:
the sriov failure is not due to this PR, its a ping that fails, and happens also on periodic
/test pull-kubevirt-e2e-kind-1.19-sriov

EdDev commented on tests/libvmi/network.go:
```diff
@@ -79,12 +79,12 @@ func InterfaceWithMac(iface *kvirtv1.Interface, macAddress string) *kvirtv1.Inte
 
 // MultusNetwork returns a Network with the given name
-func MultusNetwork(networkName string) *kvirtv1.Network {
+func MultusNetwork(nadName, networkName string) *kvirtv1.Network {
```
nit: I think the order of the arguments need to be per in reverse.
The subject here is the network, so its name should be first and only then details of its content.

Also, before this change, we assumed both will be called the same, but now it is not the case so you need to rename the existing argument to be less confusing. So how about changing it to just `name`?
The doc comment also needs to be updated.

oshoval replied:
done

EdDev reviewed (commented):
First commit review.

oshoval commented:
addressed comment
thanks

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,198 @@
+
+			orderedInterfaces := orderInterfaceStatus(vmiObj.Status.Interfaces)
+			for i := range orderedInterfaces {
+				expectedInterfaces[i].IP = orderedInterfaces[i].IP
+				expectedInterfaces[i].IPs = orderedInterfaces[i].IPs
```
I do not understand this one. You are overriding the expected interfaces with the reported ones IPs? Is this in order to normalize them for comparison?
If this is the case, it is confusing, because you involve both objects. Can you just remove the IP data from the current state and that is it?

oshoval replied:
for some reason it didnt work in the past, but now it does, done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,198 @@
+			}
+
+			Expect(reflect.DeepEqual(expectedInterfaces, orderedInterfaces)).To(BeTrue(),
+				fmt.Sprintf("interfaces doesn't match\n%v\n%v", expectedInterfaces, orderedInterfaces))
```
same, you need "do".

Regarding the assertion, I think there is a dedicated gomega one you can use instead of this DeepEqual. Then you will not need the message. In fact, `Equal` should be good enough per its description (please confirm it does work).

oshoval replied:
used ConsistOf as the next comment

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,198 @@
+				return dummyInterfaceExists(vmiObj)
+			}, 120*time.Second, 2*time.Second).Should(Equal(true))
+
+			Expect(len(vmiObj.Status.Interfaces)).To(Equal(len(expectedInterfaces)),
+				fmt.Sprintf("interfaces doesn't match\n%v\n%v", expectedInterfaces, vmiObj.Status.Interfaces))
```
I think `ConsistOf` will do the job and also cover equality, so it smells that this is not needed.

BTW, you need to use "do" and not "does" in the message (in case this one remains).

oshoval replied:
done, ConsistOf make the swap and length check redundant indeed

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,198 @@
+	}
+}
+
+// dummyInterfaceExists waits for least the dummy interface to be reported by guest agent,
+// which means there was a guest-agent report finally, and then we can compare the rest expected info.
```
- It does not wait.
- The text should not assume usage, only to describe what it does.

oshoval replied:
done

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,198 @@
+// orderInterfaceStatus orders "guest-agent" only records in predictable order
+func orderInterfaceStatus(interfaces []kvirtv1.VirtualMachineInstanceNetworkInterface) []kvirtv1.VirtualMachineInstanceNetworkInterface {
+	orderedArray := interfaces
+	length := len(orderedArray)
```
Please take care of the case where this is not the expected value. I think you expect 2 and any other number will not do what this function says it should.

I actually do not like this function at all, it is very specific but its name is very generic.
If you just want to swap places of two interfaces, just ask the question and do a simple swap.
`if ...InterfaceName != dummyInterfaceName { a, b = b, a}`

EdDev commented on tests/network/vmi_infosource.go:
```diff
@@ -0,0 +1,198 @@
+
+// orderInterfaceStatus orders "guest-agent" only records in predictable order
+func orderInterfaceStatus(interfaces []kvirtv1.VirtualMachineInstanceNetworkInterface) []kvirtv1.VirtualMachineInstanceNetworkInterface {
+	orderedArray := interfaces
```
Both point to the same list, so I do not see the point.

oshoval replied:
arrays are copied by value in golang
https://go.dev/play/p/oZN8fCbuUVn
so now the function returns a new instance

oshoval replied:
right, confused with arrays that are copied by values, but since it is a slice it isn't

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1550,6 +1553,16 @@ func getInterfaceNetworkNameByMAC(vmi *v1.VirtualMachineInstance, macAddress str
 	return ""
 }
 
+func getInterfaceInfoSourceByMAC(vmi *v1.VirtualMachineInstance, macAddress string) string {
```
I think this should be a general lookup based on mac, then used as:
`lookupInterfaceStatusByMac(vmi.Status.Interfaces, mac).InfoSource`

I guess it fits `vmispec`

oshoval replied:
done

oshoval commented:
addressed comments

oshoval commented:
/test pull-kubevirt-goveralls

saw it fails on the past + seems not related to the PR

EdDev reviewed (approved):
Thank you!

oshoval commented:
Thanks Edy

@AlonaKaplan can you please take a look ?

oshoval commented:
rebased
code from pkg/virt-handler/vm.go was moved to pkg/network/setup/netstat.go
pkg/network/setup/BUILD.bazel was updated due to that

oshoval commented:
updated the new unit tests that the rebase brought

EdDev reviewed (approved):
We have a missing test against `UpdateStatus` with `InfoSourceGuestAgent` value.
It is covered in other unit tests and on the e2e test, but it is still needed.

I am ok with adding it after the netstat_test refactoring in the following PR/s, where it is simpler to do.

oshoval commented:
rebased

oshoval commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -249,11 +250,13 @@ func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceSt
 	}
 
 	aliasesCoveredByAgent := []string{}
-	// Add alias from domain to interfaceStatus
```
Because now it adds also the InfoSource not just the alias,so it is outdated (and code is self explanatory about those 2)

oshoval commented on pkg/network/setup/netstat_test.go:
```diff
@@ -341,14 +342,14 @@ var _ = Describe("netstat", func() {
 
 		Expect(vmi.Status.Interfaces).To(Equal([]v1.VirtualMachineInstanceNetworkInterface{
-			newVMIStatusIface(networkName, nil, ifaceMAC, guestIfaceName),
+			newVMIStatusIface(networkName, nil, ifaceMAC, guestIfaceName, netvmispec.InfoSourceDomainAndGA),
```
see this please
https://github.com/kubevirt/kubevirt/pull/6793/files#diff-1c532caac99cc080ba1fb740c75c0ff712ca91509093f55cebdfaefee06ae887R287

oshoval commented on pkg/network/setup/netstat.go:
```diff
@@ -158,14 +159,18 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 					newInterface.IPs = interfaceStatus.IPs
 				}
 				delete(domainInterfaceStatusByMac, interfaceMAC)
+			} else {
```
it is true only if there is guest agent, if there isn't then we do need this else
as the whole merge won't be called

oshoval commented on pkg/network/setup/netstat.go:
```diff
@@ -158,14 +159,18 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 			// Remove the interface from domainInterfaceStatusByMac to mark it as handled
 			if interfaceStatus, exists := domainInterfaceStatusByMac[interfaceMAC]; exists {
 				newInterface.InterfaceName = interfaceStatus.InterfaceName
+				newInterface.InfoSource = interfaceStatus.InfoSource
```
Since the merge has one of 3 values:
1. domain
2. guest-agent (can be in this specific line, because we know we are iterating the spec)
3. domain and guest-agent (`InfoSourceDomainAndGA`)

then there are two options left 1 and 3, not just `InfoSourceDomainAndGA`

oshoval commented on pkg/network/setup/netstat.go:
```diff
@@ -180,6 +185,7 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 				IP:            domainInterfaceStatus.Ip,
 				IPs:           domainInterfaceStatus.IPs,
 				InterfaceName: domainInterfaceStatus.InterfaceName,
+				InfoSource:    domainInterfaceStatus.InfoSource,
```
Right, but delegated it to the one that calculate it atm (the merger), here we just copied the result

oshoval commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -249,11 +250,13 @@ func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceSt
 	for i, interfaceStatus := range interfaceStatuses {
 		if alias, exists := aliasByMac[interfaceStatus.Mac]; exists {
 			interfaceStatuses[i].Name = alias
+			interfaceStatuses[i].InfoSource = netvmispec.InfoSourceDomainAndGA
```
In order to move the whole logic to `netstat.go`
We need first this fix https://github.com/kubevirt/kubevirt/pull/6926

I tried to start with it, and saw that it uncover the leakage in case the mac is changed
https://github.com/kubevirt/kubevirt/pull/6744 (last comment shows it)

EdDev commented:
/sig network

AlonaKaplan commented:
/approve

I"m waiting for a follow-up PR removing the non guest agent information from the merge as #6744 intended to do,
And specifying the info-source directly in netstat.go (status update) and not in two plaves (the metges and the status update).

oshoval commented:
CI has some problems fetching from github, will retest soon

oshoval commented:
/retest

oshoval commented:
/retest

oshoval commented:
/retest

oshoval commented:
rebased

oshoval commented:
/test pull-kubevirt-check-tests-for-flakes

oshoval commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

for some reason CI didnt run it for 18 hours

oshoval commented:
/cherry-pick release-0.49
